File: examples/signals/src/main/csharp/org.apache.etch.examples.signals/ImplSignalsServer.cs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements. See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership. The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License. You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied. See the License for the
// specific language governing permissions and limitations
// under the License.
//

using System;

using org.apache.etch.examples.signals.types.Signals;
using Org.Apache.Etch.Bindings.Csharp.Util;
using System.Timers;

///<summary>Your custom implementation of BaseSignalsServer. Add methods here to provide
///implementation of messages from the client. </summary>
namespace org.apache.etch.examples.signals
{
  ///<summary>Implementation for ImplSignalsServer</summary>
  public class ImplSignalsServer : BaseSignalsServer
  {
    private static int clientCount;

    System.Timers.Timer signalTimer;
    private int signalCount;

    /// <summary>Constructs the ImplSignalsServer.</summary>
    /// <param name="client">a connection to the client session. Use this to
    /// send a message to the client.</param>
    public ImplSignalsServer( RemoteSignalsClient client )
      : base( client )
    {
    }

    public override string get_free_client_name()
    {
      string clientName = "client" + clientCount;

      clientCount++;

      Console.WriteLine( "Allocated client name: " + clientName );
      return clientName;
    }

    public override void _SubscribeExampleSignal( Who sender )
    {
      base._SubscribeExampleSignal( sender );
      Console.WriteLine( "Subscribed " + sender );

      if ( signalTimer == null )
      {
        Console.WriteLine( "Starting timer" );

        signalTimer = new System.Timers.Timer();
        signalTimer.Interval = 1000;
        signalTimer.Elapsed += TimerElapsedHandler;
        signalTimer.Enabled = true;
      }
    }

    public override void _UnsubscribeExampleSignal( Who sender )
    {
      base._UnsubscribeExampleSignal( sender );
      Console.WriteLine( "Unsubscribed " + sender );

      if ( client.ExampleSubscribers.Length < 1 )
      {
        Console.WriteLine( "No subscribers, stopping timer" );

        DisposeTimer();
      }
    }

    public override void _SessionNotify( object eventObj )
    {
      if ( eventObj.Equals( SessionConsts.DOWN ) )
      {
        Console.WriteLine( "Session DOWN, stopping timer" );
        DisposeTimer();
      }
    }

    private void DisposeTimer()
    {
      if ( signalTimer != null )
      {
        signalTimer.Enabled = false;
        signalTimer.Dispose();
        signalTimer = null;
      }
    }

    private void TimerElapsedHandler( object sender, ElapsedEventArgs e )
    {
      if ( signalTimer != null )
      {
        signalCount++;
        Console.WriteLine( "Sending event to " + client.ExampleSubscribers.Length + " subscribers (" + signalCount + ")" );
        client.example_signal( signalCount );
      }
    }
  }
}

File: examples/signals/src/main/java/org/apache/etch/examples/signals/ImplSignalsServer.java
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */

package org.apache.etch.examples.signals;

import java.util.Timer;
import java.util.TimerTask;

import org.apache.etch.util.core.Who;
import org.apache.etch.util.core.io.Session;

/**
 * Custom implementation of BaseSignalsServer.
 */
public class ImplSignalsServer extends BaseSignalsServer
{
	private static int clientCount = 0;
	private Timer signalTimer = null;
	private int signalCount = 0;

	/**
	 * Constructs the ImplSignalsServer.
	 *
	 * @param client a connection to the client session.
	 */
	public ImplSignalsServer( RemoteSignalsClient client )
	{
		super( client );
	}

	@Override
	public String get_free_client_name() {
		String clientName = "client" + clientCount;

		clientCount++;

		System.out.println( "Allocated client name: " + clientName );
		return clientName;
	}

	@Override
	public void _sessionNotify(Object event) throws Exception {
		if (Session.DOWN.equals(event)) {
			System.out.println( "Session DOWN, stopping timer" );
			disposeTimer();
		}
	}

	@Override
	public void _SubscribeExampleSignal(Who sender) throws Exception {
		super._SubscribeExampleSignal( sender );
		System.out.println( "Subscribed " + sender );

		if ( signalTimer == null )
		{
			System.out.println( "Starting timer" );

			signalTimer = new Timer();
			signalTimer.schedule(new NotifyClientsTask(), 1000, 1000);
		}
	}

	@Override
	public void _UnsubscribeExampleSignal(Who sender) throws Exception {
		super._UnsubscribeExampleSignal( sender );
		System.out.println( "Unsubscribed " + sender );

		if ( client._getExampleSubscribers().length < 1 )
		{
			System.out.println( "No subscribers, stopping timer" );
			disposeTimer();
		}
	}

	private void disposeTimer() {
		if (signalTimer != null) {
			signalTimer.cancel();
			signalTimer = null;
		}
	}

	class NotifyClientsTask extends TimerTask {
		@Override
		public void run() {
			signalCount++;
			System.out.println("Sending event to " + client._getExampleSubscribers().length + " subscribers (" + signalCount + ")");
			client.example_signal(signalCount);
		}
	}

}
